zerver/views/messages.py

```diff
@@ -938,33 +940,42 @@ def post_process_limited_query(rows: List[Any],
     # num_after, we want to know found_oldest and found_newest, so
     # that the clients will know that they got complete results.
 
+    if first_visible_message_id > 0:
+        visible_rows = [r for r in rows if r[0] >= first_visible_message_id]
+    else:
+        visible_rows = rows
+
+    rows_limited = len(visible_rows) != len(rows)
+
     if anchored_to_right:
         num_after = 0
-        before_rows = rows[:]
+        before_rows = visible_rows[:]
         anchor_rows = []  # type: List[Any]
         after_rows = []  # type: List[Any]
     else:
-        before_rows = [r for r in rows if r[0] < anchor]
-        anchor_rows = [r for r in rows if r[0] == anchor]
-        after_rows = [r for r in rows if r[0] > anchor]
+        before_rows = [r for r in visible_rows if r[0] < anchor]
+        anchor_rows = [r for r in visible_rows if r[0] == anchor]
+        after_rows = [r for r in visible_rows if r[0] > anchor]
 
     if num_before:
         before_rows = before_rows[-1 * num_before:]
 
     if num_after:
         after_rows = after_rows[:num_after]
 
-    rows = before_rows + anchor_rows + after_rows
+    visible_rows = before_rows + anchor_rows + after_rows
 
     found_anchor = len(anchor_rows) == 1
     found_oldest = anchored_to_left or (len(before_rows) < num_before)
     found_newest = anchored_to_right or (len(after_rows) < num_after)
+    history_limited = rows_limited and found_oldest
```

Comment:
I think this is incorrect; should `history_limited` also be true if `after_anchor = max(anchor, first_visible_message_id)` adjusted the query?

Reply:
(Is there a test that covers this corner case?)

Reply:
Yeah. Theoretically, that should be history_limited = True. But I am setting history_limited to True only if we have `found_oldest = True`. I think that is enough to show the history limited message in frontend. We show message history limited message when a user scrolls to top and found oldest but there are more messages or when users open a narrow and it has been limited. Am I missing any case?

Reply:
`rows_limited` will always be false in the case we're discussing, and so it doesn't matter what `found_oldest` is.  A better approximation would be something like this:

`(rows_limited or limited_anchor) and found_oldest` 

but I think that'd still be wrong for some rare cases where there actually aren't any messages older than the limit cutoff.  I think a correct implementation might be one that ensures we still do a `before` query, with an adjusted anchor, even when we adjust the anchor (that then gets cleared out by the `rows_limited` check).